packages/ariakit-react-core/src/focusable/focusable.tsx

```diff
@@ -343,6 +343,10 @@ export const useFocusable = createHook<TagName, FocusableOptions>(
       if (!hasFocus(element)) return;
       onFocusVisible?.(event);
       if (event.defaultPrevented) return;
+      // Make sure data-focus-visible is applied visually at the same time as
+      // other data attributes like data-active-item. See
+      // https://github.com/ariakit/ariakit/issues/4083
+      element.dataset.focusVisible = "true";
```

Comment:
I assume React will retake control of this data attribute when `data-focus-visible` is set to `undefined`, thereby removing it from the DOM, right?

Reply:
Yes, I've added an assertion to check that specifically:

https://github.com/ariakit/ariakit/blob/b4a440ed72dcaa811daa9067a29c58592996bc2f/examples/composite/test.ts#L74

But we already have other tests to ensure that. Just look for `not.toHaveAttribute("data-focus-visible")`.